### PR TITLE
EG-1266 Implementation of List type labels (Chooser, Loop, Sequence support)

### PR DESCRIPTION
**Pull request by woollett-r**

### CODE REVIEW

File: app/core/models/ocelot/Label.scala [hunk 1]
```diff
@@ -19,18 +19,61 @@ package core.models.ocelot
...
 import play.api.libs.json._
 
-case class Label(name: String, english: Option[String] = None, welsh: Option[String] = None)
+trait Label {
+  val name: String
+  val english: Option[List[String]]
+  val welsh: Option[List[String]]
+}
+
+case class ScalarLabel(name: String, english: Option[List[String]] = None, welsh: Option[List[String]] = None) extends Label
```
Comment (cmrhmrc) [inline app/core/models/ocelot/Label.scala:28]: english and welsh => List[String] as discussed

File: app/core/models/ocelot/LabelCache.scala [hunk 2]
```diff
@@ -20,31 +20,48 @@ import play.api.i18n.Lang
...
+  def updateList(name: String, english: List[String]): Labels
+  def updateList(name: String, english: List[String], welsh: List[String]): Labels
   def updatedLabels: Map[String, Label]
   def labelMap:Map[String, Label]
   def flush(): Labels
 }
 
 private class LabelCacheImpl(labels: Map[String, Label] = Map(), cache: Map[String, Label] = Map()) extends Labels {
-  def value(name: String): Option[String] = label(name).map(_.english.getOrElse(""))
+  def value(name: String): Option[String] = label(name).map{l => l.english.fold("")(c => if(c.nonEmpty) c.head else "")}
```
Comment (cmrhmrc) [inline app/core/models/ocelot/LabelCache.scala:35]: add validation label type appropriate

File: app/core/models/ocelot/LabelCache.scala [hunk 3]
```diff
@@ -20,31 +20,48 @@ import play.api.i18n.Lang
...
+  def valueAsList(name: String): Option[List[String]] = label(name).map{lbl => lbl.english.getOrElse(Nil)}
```
Comment (cmrhmrc) [inline app/core/models/ocelot/LabelCache.scala:36]: add validation label type appropriate

File: app/core/models/ocelot/LabelCache.scala [hunk 4]
```diff
@@ -20,31 +20,55 @@ import play.api.i18n.Lang
...
+  def updateList(name: String, english: List[String], welsh: List[String]): Labels
   def updatedLabels: Map[String, Label]
   def labelMap:Map[String, Label]
   def flush(): Labels
 }
 
 private class LabelCacheImpl(labels: Map[String, Label] = Map(), cache: Map[String, Label] = Map()) extends Labels {
-  def value(name: String): Option[String] = label(name).map(_.english.getOrElse(""))
+  def value(name: String): Option[String] = label(name).fold[Option[String]](None)
+    {case s:ScalarLabel => Some(if(s.english.nonEmpty)s.english.head else "") case _:ListLabel => None}
```
Comment (cmrhmrc) [inline app/core/models/ocelot/LabelCache.scala:36]: can be simplified using collect() 
e.g. `label(name).collect{case s:ScalarLabel => s.english.headOption.getOrElse("")}`

File: app/core/models/ocelot/LabelCache.scala [hunk 5]
```diff
@@ -20,31 +20,55 @@ import play.api.i18n.Lang
...
+  def valueAsList(name: String): Option[List[String]] = label(name).fold[Option[List[String]]](None)
```
Comment (cmrhmrc) [inline app/core/models/ocelot/LabelCache.scala:37]: as value() above

File: app/core/models/ocelot/LabelCache.scala [hunk 6]
```diff
@@ -20,31 +20,55 @@ import play.api.i18n.Lang
...
+  def updateList(name: String, english: List[String], welsh: List[String]): Labels = new LabelCacheImpl(labels, updateOrAddListLabel(name, english, welsh))
   def updatedLabels: Map[String, Label] = cache
   def labelMap:Map[String, Label] = labels
   def flush(): Labels = new LabelCacheImpl(labels ++ cache.toList, Map())
 
   private def label(name: String): Option[Label] = cache.get(name).fold(labels.get(name))(Some(_))
-  private def updateOrAddLabel(name: String, english: String, welsh: Option[String] = None): Map[String, Label] =
-    cache + (name -> cache.get(name).fold[Label](Label(name, Some(english), welsh))(l => Label(l.name, Some(english), welsh)))
+
+  private def updateOrAddScalarLabel(name: String, english: String): Map[String, Label] = {
```
Comment (cmrhmrc) [inline app/core/models/ocelot/LabelCache.scala:55]: think this and following private fns could be merged into one fn using default arg

File: app/core/models/ocelot/LabelCache.scala [hunk 7]
```diff
@@ -20,31 +20,55 @@ import play.api.i18n.Lang
...
+      (l => ScalarLabel(l.name, List(english))))
+  }
+
+  private def updateOrAddScalarLabel(name: String, english: String, welsh: String): Map[String, Label] = {
+    cache + (name -> cache.get(name).fold[Label]
+      (ScalarLabel(name, List(english), List(welsh)))
+      (l => ScalarLabel(l.name, List(english), List(welsh))))
+  }
+
+  private def updateOrAddListLabel(name: String, english: List[String]): Map[String, Label] =
```
Comment (cmrhmrc) [inline app/core/models/ocelot/LabelCache.scala:67]: As comment above

File: app/core/models/ocelot/LabelCache.scala (hunk 4 above)
Comment (woollett-r) [reply app/core/models/ocelot/LabelCache.scala:36]: Done

File: app/core/models/ocelot/LabelCache.scala (hunk 5 above)
Comment (woollett-r) [reply app/core/models/ocelot/LabelCache.scala:37]: Done

File: app/core/models/ocelot/LabelCache.scala (hunk 7 above)
Comment (woollett-r) [reply app/core/models/ocelot/LabelCache.scala:67]: Done

File: app/core/models/ocelot/LabelCache.scala (hunk 6 above)
Comment (woollett-r) [reply app/core/models/ocelot/LabelCache.scala:55]: Fixed

**Pull request merged**